### ISSUE TITLE
How to replace  nn.MultiheadAttention with FlashMHA when attn_mask is not None?

### ISSUE THREAD

**Issue by adoda** [HazyResearch/flash-attention#80]
Hi,
How to replace  nn.MultiheadAttention with FlashMHA when attn_mask is not None?

for example:
self.attn = nn.MultiheadAttention(d_model, n_head)

x = self.attn(x, x, x, attn_mask=attn_mask)

How to replace nn.MultiheadAttention with FlashMHA ?

**Comment by tridao** [comment]
It depends on what kind of mask. We support [key padding mask](https://github.com/HazyResearch/flash-attention/blob/45bcf37b975006fe935e5ed888e70e26ac370c90/flash_attn/flash_attention.py#L93).
However, the most performant way to deal with key_padding_mask is to remove all the padding tokens before the 1st layer, pass it through the entire model, then add back the padding tokens (called unpadding). That's how we (and others) do it for MLPerf [BERT](https://github.com/mlcommons/training_results_v2.1/blob/fb795a62ba00727da58659ad2b0cedc40fdecb8f/Azure-HazyResearch/benchmarks/bert/implementations/ND96amsr_A100_v4/modeling.py#L428).

**Comment by adoda** [comment]
How about causal attention mask?

for example:

def build_attention_mask(self):
        # lazily create causal attention mask, with full attention between the vision tokens
        # pytorch uses additive attention mask; fill with -inf
        mask = torch.empty(self.context_length, self.context_length)
        mask.fill_(float("-inf"))
        mask.triu_(1)  # zero out the lower diagonal
        return mask

**Comment by tridao** [comment]
You can pass in `causal=True`.
For performance reason, the API are different from Pytorch's API.

**Comment by itruonghai** [comment]
Hi @tridao,
I want to ask 2 questions:
1. In [this API](https://github.com/HazyResearch/flash-attention/blob/45bcf37b975006fe935e5ed888e70e26ac370c90/flash_attn/flash_attention.py#L74), the current interface only support FlashSelfAttention, can we have the MHA support for CrossAttention.
2. In the pytorch API of MultiheadAttention, the ```key_padding_mask``` and ```attn_mask``` are a bit different. Do you have plan to implement this, to make the 2 API the same? In logic, they both mean masking some part of input but with different dimension.
 
Thank you

**Comment by tridao** [comment]
CrossAttention is already supported, you can find it [here](https://github.com/HazyResearch/flash-attention/blob/main/flash_attn/modules/mha.py).

We already support sequences with different lengths (e.g., RaggedTensor/NestedTensor), which serves the same function as `key_padding_mask`. For arbitrary `attn_mask` (or bias) that's gonna take a while longer to add.

**Comment by llllly26** [comment]
But the CrossAttention is the base that covers O(n^2) complexity. Is there FlashAttention for cross-Attention that has key-padding-mask? 
thanks! @tridao

**Comment by llllly26** [comment]
> How about causal attention mask?
> 
> for example:
> 
> ```
> def build_attention_mask(self):
>         # lazily create causal attention mask, with full attention between the vision tokens
>         # pytorch uses additive attention mask; fill with -inf
>         mask = torch.empty(self.context_length, self.context_length)
>         mask.fill_(float("-inf"))
>         mask.triu_(1)  # zero out the lower diagonal
>         return mask
> ```

Hi, have you solved the problem? Is `causal attention mask`  successfully work? 
thanks! @adoda
